phoenix-core/src/test/java/org/apache/phoenix/query/QueryServicesHelperTest.java

```diff
@@ -0,0 +1,205 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.phoenix.query;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.phoenix.exception.PhoenixNonRetryableRuntimeException;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class QueryServicesHelperTest {
+
+    static GuidePostsCache testCache = null;
+
+    public static class TestGuidePostsCacheFactory implements  GuidePostsCacheFactory {
+
+        public static volatile int count=0;
+
+        public TestGuidePostsCacheFactory() {
+            count++;
+        }
+
+        @Override
+        public GuidePostsCache getGuidePostsCacheInterface(ConnectionQueryServices queryServices,
+                Configuration config) {
+            return testCache;
+        }
+    }
+
+
+
+    public static CountDownLatch latch1 ;
+    public static CountDownLatch latchCollect ;
+    ConcurrentMap<GuidePostsCacheFactory,Object> concurrentMap;
+
+
+    public static class TestGuidePostsLatchedCacheFactory implements  GuidePostsCacheFactory {
+        public TestGuidePostsLatchedCacheFactory() {
+            try {
+                latch1.await();
```

Comment:
Can we add a timeout here? In rare cases, we don't want to get stuck out here and cause build jobs to completely fail.

Reply:
Done/Refactored.